subiquity/tests/api/test_api.py

```diff
@@ -0,0 +1,521 @@
+    need_spawn_server = True
+
+    async def get(self, query, **kwargs):
+        return await self.request('GET', query, **kwargs)
+
+    async def post(self, query, data=None, **kwargs):
+        return await self.request('POST', query, data, **kwargs)
+
+    async def request(self, method, query, data=None, **kwargs):
+        params = {}
+        for key in kwargs:
+            params[key] = dumps(kwargs[key])
+        data = dumps(data)
+        info = f'{method} {query}'
+        if params:
+            for i, key in enumerate(params):
+                joiner = '?' if i == 0 else '&'
+                info += f'{joiner}{key}={params[key]}'
+        print(info)
+        async with self.session.request(method, f'http://a{query}',
+                                        data=data, params=params) as resp:
+            resp.raise_for_status()
+            content = await resp.content.read()
+            return loads(content.decode())
+
+    async def server_startup(self):
+        for _ in range(20):
+            try:
+                await self.get('/meta/status')
+                return
+            except aiohttp.client_exceptions.ClientConnectorError:
+                await asyncio.sleep(.5)
+        raise Exception('timeout on server startup')
+
+    async def server_shutdown(self, immediate=True):
+        try:
+            await self.post('/shutdown', mode='POWEROFF', immediate=immediate)
+            raise Exception('expected ServerDisconnectedError')
+        except aiohttp.client_exceptions.ServerDisconnectedError:
+            return
+
+    async def spawn_server(self):
+        env = os.environ.copy()
+        env['SUBIQUITY_REPLAY_TIMESCALE'] = '100'
+        cmd = 'python3 -m subiquity.cmd.server --dry-run --bootloader uefi' \
+              + ' --machine-config ' + self.machine_config
+        cmd = cmd.split(' ')
+        self.proc = await astart_command(cmd, env=env)
+        self.server = asyncio.create_task(self.proc.communicate())
+
+    async def asyncSetUp(self):
+        if self.need_spawn_server:
+            await self.spawn_server()
+
+        # setup client
+        conn = aiohttp.UnixConnector(path=socket_path)
+        self.session = aiohttp.ClientSession(connector=conn)
+        await self.server_startup()
+
+    async def asyncTearDown(self):
+        if self.need_spawn_server:
+            await self.server_shutdown()
+            await self.session.close()
+            await self.server
+            try:
+                self.proc.kill()
+            except ProcessLookupError:
+                pass
+
+
+# class TestDebug(TestAPI):
+#     need_spawn_server = False
+
+#     @unittest.skip("useful for interactive debug only")
+#     async def test_v2_delete_debug(self):
+#         await self.get('/storage/v2')
+
+
+class TestSimple(TestAPI):
+    @timeout(5)
+    async def test_not_bitlocker(self):
+        resp = await self.get('/storage/has_bitlocker')
+        self.assertEqual(0, len(resp))
```

Comment:
The problem with implicit set up is that you have to know what it does to understand that this test is correct (not only do you have to know that a server has been implicitly started, you have to know that it uses a config that doesn't have any bitlocker disks!). It's more verbose but I would prefer this test to be written more like:

```
async def test_not_bitlocker(self):
    server = await self.startServerForConfig('examples/simple.json')
    resp = await server.get('/storage/has_bitlocker')
    self.assertEqual(0, len(resp))
```

This would also mean that you could have the "test_not_bitlocker" and "test_yes_bitlocker" tests next to each other.